master/buildbot/process/buildstep.py

```diff
@@ -1086,7 +1086,10 @@ def getText(self, cmd, results):
             return self.describe(True) + ["failed"]
 
     def getText2(self, cmd, results):
-        return ["Step \"%s\"" % self.name]
+        optional = ""
+        if not self.haltOnFailure and self.flunkOnFailure:
```

Comment:
Shouldn't this be `if not self.haltOnFailure and not self.flunkOnFailure`?

In fact, I think we can ignore `haltOnFailure`, AFAIK it just ends the build, but doesn't say anything about failing or not. So the condition could be just `if not self.flunkOnFailure`

Reply:
from katana master docs (this very long html docs about buildbot)
haltOnFailure: if True, a FAILURE of this build step will cause the build to halt immediately.

In my opinion, when `haltOnFailure==True` step breaks build, so we cannot add "Optional " status. So we have to check `not self.haltOnFailure`

flunkOnFailure: when True, a FAILURE of this build step will mark the overall build as a FAILURE. The remaining steps will still be executed.

And we have two cases (with haltOnFailure==False)
`flunkOnFailure==False`, step FAILED but "who cares?" build is fine and do next steps
`flunkOnFailure==True`, ok, we mark build as Failed and build do next steps

Hmmm.. when I'm thinking again about it now, maybe we should only check `not haltOnFailure`. flunkOfFailure in both cases makes step as "Optionaly". What fou you thing?

Reply:
Step conditions:

`flunk=True, halt=True` a step failure will fail the whole build, and stop immediately (NOT an optional step)

`flunk=True, halt=False` a step failure will fail the whole build, but continue running it (NOT an optional step, though I don't think there are any cases like this)

`flunk=False, halt=True` a step failure will not fail the whole build, but it will stop immediately (OPTIONAL step, though I don't think there are any cases like this)

`flunk=False, halt=false` a step failure will not fail the whole build, and it will continue running (OPTIONAL step)

Reply:
flunk is the only thing that has affects whether that step is optional or not, since technically halt only refers to stopping the build or not (doesn't say anything about status).

In practice it makes little difference, since I believe that they are always the same (both True or both False)

Reply:
You convinced me :-)